### PR TITLE
Sort WorkerPool in executions

### PR DESCRIPTION
**Pull request by onderkalaci**
We don't need to backport this (or  maybe even merge?), because we're already implicitly enforcing this in the executor/planner. This came up in the context of #3849 where COPY doesn't have any order while opening connections. So, we could end up with this [situation](https://github.com/citusdata/citus/pull/3849#pullrequestreview-431792473).

```
We sort the workerList because adaptive connection management
(e.g., OPTIONAL_CONNECTION) requires any concurrent executions
to wait for the connections in the same order to prevent any
starvation. If we don't sort, we might end up with:
      Execution 1: Get connection for worker 1, wait for worker 2
      Execution 2: Get connection for worker 2, wait for worker 1

and, none could proceed. Instead, we enforce every execution establish
the required connections to workers in the same order.
```

TODO:
- [ ] We need to make sure whether COPY can enforce the same ordering. Otherwise, merging this doesn't help much

### CODE REVIEW

File: src/backend/distributed/executor/adaptive_executor.c [hunk 1]
```diff
@@ -1923,6 +1937,31 @@ AssignTasksToConnectionsOrWorkerPool(DistributedExecution *execution)
...
+ * The function compares two worker nodes by their host name and port
+ * number.
+ */
+static int
+WorkerPoolCompare(const void *lhsKey, const void *rhsKey)
+{
+	const WorkerPool *workerLhs = *(const WorkerPool **) lhsKey;
+	const WorkerPool *workerRhs = *(const WorkerPool **) rhsKey;
+
+	int nameCompare = strncmp(workerLhs->nodeName, workerRhs->nodeName,
```
Comment (marcocitus) [inline src/backend/distributed/executor/adaptive_executor.c:1952]: Maybe just sort by node ID?

Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:1952]: I remember that we removed  `FindWorkerNode` calls in the past, do we want to bring them  back? https://github.com/citusdata/citus/pull/2848/commits/a3811b1e5517b8c0c5ae3a25f8753c4d9768331b

It seems that the only `findWorkerNode` was the one that we called where we need now.

Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:1952]: I might do some benchmarks before deciding on this

Comment (onderkalaci) [reply src/backend/distributed/executor/adaptive_executor.c:1952]: It turns out that calling `FindWorkerNode` is pretty expensive compared to this sorting
```
worker_name_length | worker count | sort by name + compare by name | sort by id + compare by id
-- | -- | -- | --
15 | 3 | 0 microseconds | 3 microseconds
15 | 5 | 1 microseconds | 3 microseconds
15 | 10 | 2 microseconds | 4 microseconds
15 | 20 | 2 microseconds | 20 microseconds
15 | 50 | 5 microseconds | 60 microseconds
15 | 150 | 15 microseconds | 228 microseconds
15 | 500 | 62 microseconds | 755 microseconds
50 | 3 | 0 microseconds | 3 microseconds
50 | 5 | 1 microseconds | 3 microseconds
50 | 10 | 2 microseconds | 4 microseconds
50 | 20 | 2 microseconds | 20 microseconds
50 | 50 | 5 microseconds | 60 microseconds
50 | 150 | 16 microseconds | 228 microseconds
50 | 500 | 58 microseconds | 755 microseconds
200 | 3 | 0 microseconds | 3 microseconds
200 | 5 | 1 microseconds | 3 microseconds
200 | 10 | 2 microseconds | 4 microseconds
200 | 20 | 2 microseconds | 20 microseconds
200 | 50 | 4 microseconds | 60 microseconds
200 | 150 | 14 microseconds | 228 microseconds
200 | 500 | 64 microseconds | 755 microseconds
```

worker_name_length: means only the last characters differ in the workerNames, so strncmp goes to the end

Comment (marcocitus) [reply src/backend/distributed/executor/adaptive_executor.c:1952]: I guess we would have to add the node ID to the shard placement, but perhaps that's probably overengineering this a bit.

**Comment by onderkalaci** [comment]
>  We need to make sure whether COPY can enforce the same ordering. Otherwise, merging this doesn't help much


We decided that it is still useful to explicitly sort the workers in the executor anyway. We're currently relying on planner to sort implicitly based on shardid/activeshardplacements which actually means we always have a consistent worker order for competing tasks. Still, sorting is cheap and we explicitly sort even just for executor for now

**Pull request merged**